test/builtin/test_assignment.py

```diff
@@ -238,3 +238,7 @@ def test_predecrement():
     check_evaluation(
         "--5", "4", failure_message="Set::setraw: Cannot assign to raw object 5."
     )
+
+
+def test_assign_list():
+    check_evaluation("G[x_Real]=x^2; a={G[x]}; {x=1.; a, x=.; a}", "{{1.}, {G[x]}}")
```

Comment:
The idea of this test was to check that the stored list in ``a`` is not altered during the evaluation. Curiously, we pass this test with the current implementation of ``ListExpression.rewrite_apply_eval_step``. This happens because during the regular evaluation, the ``ListExpression`` stored in ``a`` is converted back into an ``Expression`` in the ``apply_rule`` call.

Reply:
Thanks!